woocommerce/woocommerce-gateway-stripe
# Display time of last webhook in settings

Fixes #1200

#### Changes proposed in this Pull Request:

Appends a message to the webhook settings description, with the date and time of the most recently received webhook event.

(Also tweaks the order of the settings, to group webhook-related items.)

<img width="788" src="https://user-images.githubusercontent.com/1867547/83553899-f005ec80-a4d9-11ea-94b1-831a789a42bf.png">

## Testing instructions

Unit tests verify test and live mode messages. You can test manually using either test or live mode:

- Use stripe cli to send valid test webhooks: `stripe listen --forward-to "localhost:8082/?wc-api=wc_stripe"`.
- Paste your webhook secret in settings.

**Case 1: (Nominal case): Most recent webhook successful:**

- Run `stripe trigger payment_intent.created` to simulate a successful webhook and notice the following message:

> The most recent [mode] webhook, timestamped [timestamp], was processed successfully.

**Case 2: No webhook ever received (since this version of the plugin was installed):**

- Delete previous webhook options in the database: `DELETE FROM wp_options WHERE option_name LIKE 'wc_stripe_wh_%';`
- Notice the following message:

> No [mode] webhooks have been received since monitoring began at [timestamp].

**Case 3: Most recent webhook unsuccessful with a prior success:**

- Run `stripe trigger payment_intent.created`
- Run `curl -X POST "http://localhost:8082/?wc-api=wc_stripe"`
- Notice the following message:

> Warning: The most recent [mode] webhook, received at [timestamp], could not be processed. Reason: The webhook was missing expected body. (The last [mode] webhook to process successfully was timestamped [timestamp].)

**Case 4: Most recent webhook unsuccessful with no prior successes:**

- Delete previous webhook options in the database: `DELETE FROM wp_options WHERE option_name LIKE 'wc_stripe_wh_%';`
- Run `curl -X POST "http://localhost:8082/?wc-api=wc_stripe"`
- Notice the following message:

> Warning: The most recent [mode] webhook, received at [timestamp], could not be processed. Reason: The webhook was missing expected body. (No [mode] webhooks have been processed successfully since monitoring began at [timestamp].)

-------------------
